### PR TITLE
Hetero shared model

### PR DESCRIPTION
**Pull request by eugenevinitsky**
## What do these changes do?

1. Fixed a bug in shared model that was causing the models to not actually be shared. 
2. Implemented heterogeneous shared models, to enable mixed traffic lights/AV environments. 
3. Pulled in upstream ray changes. This makes it hard to read the request so, the important file is multiagent_fcnet

### CODE REVIEW

File: python/ray/rllib/models/multiagentfcnet.py [hunk 1]
```diff
@@ -37,10 +37,22 @@ def _init(self, inputs, num_outputs, options):
 
         # check for a shared model
         shared_model = custom_options.get("multiagent_shared_model", 0)
+        shared_model = shared_model or isinstance(shared_model, list) # for multiple shared models
+        shared_model_list = custom_options.get("multiagent_shared_model", 0) # how many models should be shared
+        shared_model_list = shared_model_list if isinstance(shared_model_list, list) else [len(hiddens)]
+
         reuse = tf.AUTO_REUSE if shared_model else False
         outputs = []
+        model_counter = 0
```
Comment (cathywu) [inline python/ray/rllib/models/multiagentfcnet.py]: Add a comment for what model_counter and scope_counter are.

File: python/ray/rllib/models/multiagentfcnet.py [hunk 2]
```diff
@@ -37,10 +37,22 @@ def _init(self, inputs, num_outputs, options):
 
         # check for a shared model
         shared_model = custom_options.get("multiagent_shared_model", 0)
+        shared_model = shared_model or isinstance(shared_model, list) # for multiple shared models
+        shared_model_list = custom_options.get("multiagent_shared_model", 0) # how many models should be shared
```
Comment (cathywu) [inline python/ray/rllib/models/multiagentfcnet.py]: This is a bit confusing. The variable name implies that this is a list, but the default value is an integer and the comment also suggests that this is an integer.

File: python/ray/rllib/models/multiagentfcnet.py [hunk 3]
```diff
@@ -37,10 +37,22 @@ def _init(self, inputs, num_outputs, options):
 
         # check for a shared model
         shared_model = custom_options.get("multiagent_shared_model", 0)
+        shared_model = shared_model or isinstance(shared_model, list) # for multiple shared models
```
Comment (cathywu) [inline python/ray/rllib/models/multiagentfcnet.py]: "multiple shared models" = heterogeneous models?

**Review by cathywu — COMMENTED** [review]
I left a few clarifying questions, I'm not entirely sure what these changes are doing.

File: python/ray/rllib/models/multiagentfcnet.py (hunk 3 above)
Comment (eugenevinitsky) [reply python/ray/rllib/models/multiagentfcnet.py]: Yeah

File: python/ray/rllib/models/multiagentfcnet.py [hunk 4]
```diff
@@ -37,10 +37,24 @@ def _init(self, inputs, num_outputs, options):
 
         # check for a shared model
         shared_model = custom_options.get("multiagent_shared_model", 0)
+        shared_model = shared_model or isinstance(shared_model, list) # for heterogeneous shared models
+        # the list contains how many agents share each model
+        shared_model_list = custom_options.get("multiagent_shared_model", [len(hiddens)])
+
         reuse = tf.AUTO_REUSE if shared_model else False
         outputs = []
+        # keeps track of how many models we have set shared so far
```
Comment (cathywu) [inline python/ray/rllib/models/multiagentfcnet.py]: "set shared" --> "set as shared"?

**Review by cathywu — APPROVED** [review]
Does this assume that the first k1 agents share a model, the next k2 agents share a different model, etc.?

**Comment by cathywu** [comment]
If so, that should be written somewhere.

File: python/ray/rllib/models/multiagentfcnet.py (hunk 4 above)
Comment (eugenevinitsky) [reply python/ray/rllib/models/multiagentfcnet.py]: Yes, this is the intent of the request. I'll modify that

**Pull request merged**